crates/metrics/src/future.rs

```diff
@@ -1,129 +1,211 @@
+pub trait FutureExt: Sized {
+    /// Consumes the future, returning a new future that records the executiion
+    /// metrics of the inner future.
+    fn with_metrics(self, metrics: impl Into<&'static Metrics>) -> Metered<Self> {
+        Metered::new(self, metrics)
     }
 }
 
+impl<F> FutureExt for F where F: Future {}
+
+/// [`Future`] wrapper collecting [`Metrics`] of inner [`Future`] `F`.
 #[pin_project::pin_project]
 #[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct TaskMetricsFuture<F, R>
-where
-    F: Future,
-    R: TaskMetricsRecorder,
-{
+pub struct Metered<F> {
     #[pin]
-    inner: F,
-    stats: Stats<R>,
+    future: F,
+    state: State,
 }
 
-impl<F, R> TaskMetricsFuture<F, R>
-where
-    F: Future,
-    R: TaskMetricsRecorder,
-{
-    pub fn new(inner: F, recorder: R) -> Self {
+struct State {
+    started_at: Option<Instant>,
+    is_finished: bool,
+
+    poll_duration_sum: Duration,
+    poll_duration_max: Duration,
+    polls_count: usize,
+
+    metrics: &'static Metrics,
+}
+
+impl<F> Metered<F> {
+    fn new(future: F, metrics: impl Into<&'static Metrics>) -> Self {
+        let metrics = metrics.into();
+
+        metrics.created.increment(1);
+
         Self {
-            inner,
-            stats: Stats::new(recorder),
+            future,
+            state: State {
+                started_at: None,
+                is_finished: false,
+                poll_duration_sum: Duration::from_secs(0),
+                poll_duration_max: Duration::from_secs(0),
+                polls_count: 0,
+                metrics,
+            },
         }
     }
 }
 
-impl<F, R> Future for TaskMetricsFuture<F, R>
-where
-    F: Future,
-    R: TaskMetricsRecorder,
-{
+impl From<&'static Lazy<Metrics>> for &'static Metrics {
+    fn from(lazy: &'static Lazy<Metrics>) -> Self {
+        lazy.get_or_register()
+    }
+}
+
+impl<F: Future> Future for Metered<F> {
     type Output = F::Output;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        let poll_start = Instant::now();
-        let this = self.project();
-        let result = this.inner.poll(cx);
+        let mut this = self.project();
+        let state = &mut this.state;
 
-        if result.is_ready() {
-            this.stats.completed = true;
+        if state.started_at.is_none() {
+            state.started_at = Some(Instant::now());
+            state.metrics.started.increment(1);
         }
 
-        this.stats.poll_entries += 1;
-        this.stats.poll_duration += poll_start.elapsed();
+        let poll_started_at = Instant::now();
+        let result = this.future.poll(cx);
+        let poll_duration = poll_started_at.elapsed();
+
+        state.poll_duration_sum += poll_duration;
+        state.poll_duration_max = state.poll_duration_max.max(poll_duration);
+        state.polls_count += 1;
+
+        if result.is_ready() && !state.is_finished {
```

Comment:
You're mixing fused and regular future approaches here. I'm not sure if it makes sense to check `state.is_finished` flag here, unless you're wrapping the inner future in `Fuse`, and implement `FusedFuture` fpr `Metered`.

Reply:
I haven't change this part of the code, just moved it. But I think the original idea was to handle cases when the underlying futur is `Fused` indeed

Reply:
Fused futures should implement `FusedFuture`. Doing polls on a `ready` future is UB, so it doesn't make sense to check this flag anyway. I would recommend removing the check here, and assume nothing is fused there (i.e. the inner future). And then implement `FusedFuture` for `Metered` based on the `state.is_finished` flag. This way the poll is cleaner in terms of logic, and it'd be properly fused.